src/v/utils/xid.cc

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "utils/xid.h"
+
+#include "bytes/iobuf_parser.h"
+#include "serde/rw/array.h"
+#include "serde/rw/rw.h"
+#include "ssx/sformat.h"
+
+#include <fmt/format.h>
+
+namespace {
+static constexpr std::array<char, 32> b32_alphabet{
+  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a',
+  'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
+  'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v'};
+
+using decoder_t = std::array<uint8_t, 256>;
+
+constexpr decoder_t build_decoder_table() {
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    std::array<uint8_t, 256> ret = {0xff};
+
+    for (uint8_t i = 0; i < b32_alphabet.size(); ++i) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        ret[b32_alphabet[i]] = i;
+    }
+    return ret;
+}
+
+static constexpr decoder_t decoder = build_decoder_table();
+} // namespace
+
+invalid_xid::invalid_xid(const ss::sstring& current_string)
+  : _msg(ssx::sformat("String '{}' is not a valid xid", current_string)) {}
+
+xid xid::from_string(const ss::sstring& str) {
+    if (str.size() != str_size) {
+        throw invalid_xid(str);
```

Comment:
I ran the test with coverage enabled, and I don't think this is in the test coverage.